### PR TITLE
Let `kotlin-dsl` plugin disable Kotlin compilation avoidance

### PR DESCRIPTION
**Pull request by eskatos**
* Fixes https://github.com/gradle/gradle/issues/24815
* Caused by https://youtrack.jetbrains.com/issue/KT-57757
* Will be reenabled in https://github.com/gradle/gradle/issues/24816

----

This PR upgrades the `kotlin-dsl` plugin, including the version used in the `gradle/gradle` build.

As a consequence building `gradle/gradle` emits the following warning:
```
This version of Gradle expects version '4.0.11' of the `kotlin-dsl` plugin but version '4.0.13' has been applied to project ':build-logic:profiling'.
Let Gradle control the version of `kotlin-dsl` by removing any explicit `kotlin-dsl` version constraints from your build logic.
```

This warning will disappear once we update the wrapper with a new nightly.

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/base/KotlinDslBasePlugin.kt [hunk 1]
```diff
@@ -47,11 +47,18 @@ abstract class KotlinDslBasePlugin : Plugin<Project> {
 
         apply<KotlinDslCompilerPlugins>()
         addGradleKotlinDslDependencyTo(kotlinArtifactConfigurationNames)
+
+        disableKotlinCompilationAvoidance()
```
Comment (asodja) [inline subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/base/KotlinDslBasePlugin.kt:51]: 🤔 Should we move the line before we apply the Kotlin plugin? Although probably plugin doesn't read that property on apply

**Review by asodja — APPROVED** [review]
Added one question about the order of applying KGP and setting the property in the plugin.

After we merge that and update wrapper we can also remove these workaround, right?
https://github.com/gradle/gradle/blob/af8cc6b4e24537ef9c0a308dea5baaeb0c4ea055/build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts#L227-L231

File: subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/base/KotlinDslBasePlugin.kt (hunk 1 above)
Comment (eskatos) [reply subprojects/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/base/KotlinDslBasePlugin.kt]: :+1: better safe than sorry

**Comment by eskatos** [comment]
Yes removing the workarounds will need the wrapper update first.

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**